addons/stock/models/stock_location.py

```diff
@@ -136,6 +143,31 @@ def _compute_warehouse_id(self):
                     loc.warehouse_id = view_by_wh[view_location_id]
                     break
 
+    @api.depends('child_ids.usage', 'child_ids.child_internal_location_ids')
+    def _compute_child_internal_location_ids(self):
```

Comment:
It is easier and more performant to use the parent_path technical fields (like for the `_compute_warehouse_id`). Making union of record set in a double loop has a complexity of N^3 (not good).
A suggestion:

```python
all_children = self.search([('id', 'child_of', self.ids), ('usage', '=', 'internal')])
all_children.read(['parent_path'])
for loc in self:
    loc.child_internal_location_ids = all_children.filtered(lambda l_c: l_c.parent_path.startswith(loc.parent_path))
```

The complexity can still be a issue (N²) if there is a lot of children location (maybe we can do better). What do you think ?

Reply:
More precisely, complexity of my first implementation is `M*N*N`, where M is `len(self)`, N is `len(all_children)`.

Your drops complexity to M*N by using `parent_path`. It could be further improved to `M + N` by sorting by `parent_path` and using two parallel iterators:

```py
        # fetch all children in batch (remember, that 'child_of' matches parents too)
        all_children = self.search([('id', 'child_of', self.ids), ('usage', '=', 'internal')], order='parent_path')

        # prefetch parent information
        all_children.read(['parent_path'])

        # sorted(['1/2/3/', '1/', '1/2/'], reverse=True) ->  ['1/2/3/', '1/2/', '1/']

        self_sorted = self.sorted('parent_path', reverse=True)

        self_iter = (loc for loc in self_sorted)

        # Each child may have few ancestors, e.g. 1/2/3/ is child of 1/, 1/2/
        # First, find the most closed ancestors
        for child in all_children:
            for parent in self_iter:
                parent.child_internal_location_ids = self.browse()
                if child.parent_path.startswith(parent.parent_path):
                    parent.child_internal_location_ids |= child
                    break

        # Now find ancestors in self
        self_iter = (loc for loc in self_sorted)
        prev_loc = next(self_iter)
        while True:
            loc = next(self_iter, None)
            if loc is None:
                break
            if prev_loc.parent_path.startswith(loc):
                loc.child_internal_location_ids |= prev_loc.child_internal_location_ids
            prev_loc = loc

```
Do you want proceed with this solution?

Reply:
I made some tests:
There is issue with the recursion depends (my bad), the `recursive=True` avoid any prefetch, then the batched version of the compute is useless (`_compute_warehouse_id` is useless batched too) two Solution :
    - To keep the prefetch and batched version: remove the `recursive=True` AND (readd your previous invalidate_cache OR we don't care about depends/invalidate_cached because we never change location data in the same transaction that when we read `the child_internal_location_ids`)
    - Keep the recursive depends with the naive compute (the search by location) and we don't need to add with_prefetch()

IMO, we should keep the naive solution (small diff, easy solution, avoid complicated compute) and we don't care about prefetch because recursive depends deactivate it.
What do you think ?

Reply:
You mean that `self` is always singleton for `recursive=True`?

I've updated the PR